### PR TITLE
Batching support for watchers ⚖️📈

### PR DESCRIPTION
**Pull request by mikegrima**
- Currently supporting IAMRole watcher as a POC.

This is a POC for batching watchers. This is to hopefully alleviate issues with too many files being open for very large workloads with thousands of a particular item.

### CODE REVIEW

**Comment by aebie** [comment]
See issue https://github.com/Netflix/security_monkey/issues/556

**Comment by mikegrima** [comment]
This should be stable now for IAM Roles.  I would like to add in some tests, however.

**Comment by mikegrima** [comment]
Need to add 2 more unit tests and this should be good to go.

**Comment by mikegrima** [comment]
Should be ready?

File: security_monkey/watcher.py [hunk 1]
```diff
@@ -295,16 +312,63 @@ def find_modified(self, previous=[], current=[], exception_map={}):
...
+            self.find_new(previous=prev, current=current)
+            self.find_modified(previous=prev, current=current, exception_map=exception_map)
+
+    def find_changes_batch(self, items, exception_map):
+        # Given the list of items, find new items that don't yet exist:
+        durable_items = []
+
+        # Need to pull out the technology, and the account from the database for
+        # proper queryings...
+        account = datastore.Account.query.filter(Account.name == self.accounts[0]).scalar()
```
Comment (mikegrima) [inline security_monkey/watcher.py]: Replace `scalar()` with `one()` since if technology and account are not found -- that is a problem.

Comment (mikegrima) [inline security_monkey/watcher.py]: Makes an assumption that there is only one account. -- Need to have logic for tracking what the current account is.

File: security_monkey/watcher.py [hunk 2]
```diff
@@ -295,16 +312,63 @@ def find_modified(self, previous=[], current=[], exception_map={}):
...
+
+            persist_item(item, db_item, technology, account, complete_hash, durable_hash, is_durable)
+
+            if is_durable:
+                durable_items.append(item)
+
+        return durable_items
+
+    def find_deleted_batch(self, exception_map):
+        account = datastore.Account.query.filter(Account.name == self.accounts[0]).scalar()
```
Comment (mikegrima) [inline security_monkey/watcher.py]: `scalar()` and tracking of the current account.

File: security_monkey/watcher.py [hunk 3]
```diff
@@ -295,16 +312,63 @@ def find_modified(self, previous=[], current=[], exception_map={}):
...
+        technology = datastore.Technology.query.filter(Technology.name == self.index).scalar()
+        arns = [item["Arn"] for item in self.total_list]
+
+        from datastore_utils import delete_old_revisions
+        return delete_old_revisions(self, arns, account, technology)
```
Comment (mikegrima) [inline security_monkey/watcher.py]: Rename to `inactivate_old_revisions` for clarity.

File: security_monkey/watchers/iam/iam_role.py [hunk 4]
```diff
@@ -35,34 +35,68 @@ class IAMRole(Watcher):
     def __init__(self, accounts=None, debug=False):
         super(IAMRole, self).__init__(accounts=accounts, debug=debug)
 
+        self.batched_size = 200
+        self.done_slurping = False
+
```
Comment (mikegrima) [inline security_monkey/watchers/iam/iam_role.py]: Delete this space.

File: security_monkey/watchers/iam/iam_role.py [hunk 5]
```diff
@@ -35,34 +35,68 @@ class IAMRole(Watcher):
     def __init__(self, accounts=None, debug=False):
         super(IAMRole, self).__init__(accounts=accounts, debug=debug)
 
+        self.batched_size = 200
```
Comment (mikegrima) [inline security_monkey/watchers/iam/iam_role.py]: Set to 100.

File: security_monkey/watchers/iam/iam_role.py [hunk 6]
```diff
@@ -35,34 +35,68 @@ class IAMRole(Watcher):
...
+        return get_role(dict(role), **kwargs)   # Need to send a copy, since we don't want to alter the total list!
 
-    def slurp(self):
+    def slurp_list(self):
         self.prep_for_slurp()
 
         @iter_account_region(index=self.index, accounts=self.accounts, exception_record_region='universal')
+        def get_role_list(**kwargs):
+            # Need to get the initial roles if we haven't already:
+            if len(self.total_list) == 0:
```
Comment (mikegrima) [inline security_monkey/watchers/iam/iam_role.py]: Is this necessary?

File: security_monkey/watchers/iam/iam_role.py [hunk 7]
```diff
@@ -35,34 +35,68 @@ class IAMRole(Watcher):
...
+                app.logger.debug("Fetching the full list of {index} that need to be slurped from {account}...".format(
+                    index=self.i_am_plural, account=self.accounts[0]
+                ))
+                self.list_roles(**kwargs)
+
+                # are there any roles?
+                if len(self.total_list) == 0:
+                    self.done_slurping = True
+
+                return [], kwargs.get('exception_map', {})
```
Comment (mikegrima) [inline security_monkey/watchers/iam/iam_role.py]: Does not need to return a list anymore. Should just return the `exception_map`.

File: security_monkey/watchers/iam/iam_role.py [hunk 8]
```diff
@@ -35,34 +35,68 @@ class IAMRole(Watcher):
     def __init__(self, accounts=None, debug=False):
         super(IAMRole, self).__init__(accounts=accounts, debug=debug)
 
+        self.batched_size = 200
+        self.done_slurping = False
+
+        self.next_role = 0
```
Comment (mikegrima) [inline security_monkey/watchers/iam/iam_role.py]: Not using this -- delete it.

File: security_monkey/watchers/iam/iam_role.py [hunk 9]
```diff
@@ -35,34 +35,68 @@ class IAMRole(Watcher):
...
-            roles = self.list_roles(**kwargs)
-
-            for role in roles:
-                role = self.process_role(role, name=role['RoleName'], **kwargs)
+            role_counter = self.batch_counter * self.batched_size
+            while self.batched_size - len(item_list) > 0 and not self.done_slurping:
+                # The next role is going to be calculated as follows:
+                current_role = self.total_list[role_counter]
+                role = self.process_role(current_role, name=current_role["RoleName"], **kwargs)
                 if role:
```
Comment (mikegrima) [inline security_monkey/watchers/iam/iam_role.py]: Move the exception comment here.

File: security_monkey/scheduler.py [hunk 10]
```diff
@@ -38,24 +38,61 @@ def run_change_reporter(account_names, interval=None):
...
+            else:
+                # Just fetch normally...
+                _slurp_delta_save(cw)
+
+    # Batched monitors have already been monitored, and they will be skipped over.
     audit_changes(accounts, monitor_names, False, debug)
     db.session.close()
 
 
+def _slurp_delta_save(cw):
```
Comment (mikegrima) [inline security_monkey/scheduler.py]: Move this back out of a function.

File: security_monkey/scheduler.py [hunk 11]
```diff
@@ -38,24 +38,61 @@ def run_change_reporter(account_names, interval=None):
...
+                    (items, exception_map) = cw.slurp()
+
+                    audit_items.extend(cw.find_changes(current=items, exception_map=exception_map))
+
+                # Delete the items that no longer exist:
+                app.logger.debug("Deleting all items for {technology}/{account} that no longer exist.".format(
+                    technology=cw.i_am_plural, account=account_name
+                ))
+                cw.find_deleted_batch(account_name)
+                _audit_specific_changes(mon, audit_items, False, debug)
```
Comment (mikegrima) [inline security_monkey/scheduler.py]: Need to test what exceptions do to the flow -- does it contain it or does it just kill the whole thing?

File: security_monkey/scheduler.py [hunk 12]
```diff
@@ -38,24 +38,61 @@ def run_change_reporter(account_names, interval=None):
...
         for mon in monitors:
             cw = mon.watcher
-            (items, exception_map) = cw.slurp()
-            cw.find_changes(current=items, exception_map=exception_map)
-            cw.save()
+            if mon.batch_support:
+                audit_items = []
+
+                # Fetch the full list of items that we need to obtain:
+                _, exception_map = cw.slurp_list()
```
Comment (mikegrima) [inline security_monkey/scheduler.py]: Remove the `_`, since it won't return a tuple.

File: security_monkey/scheduler.py [hunk 13]
```diff
@@ -38,24 +38,61 @@ def run_change_reporter(account_names, interval=None):
...
+                if len(exception_map) > 0:
```
Comment (mikegrima) [inline security_monkey/scheduler.py]: Add region logic to the output.

File: security_monkey/scheduler.py [hunk 14]
```diff
@@ -38,24 +38,61 @@ def run_change_reporter(account_names, interval=None):
...
+                    app.logger.error("Exceptions have caused nothing to be fetched for {technology}/{account}..."
+                                     " CANNOT CONTINUE FOR THIS WATCHER!".format(technology=cw.i_am_plural,
+                                                                                 account=account_name))
+                    continue
+
+                while not cw.done_slurping:
+                    app.logger.debug("Fetching a batch of {batch} items for {technology}/{account}.".format(
```
Comment (mikegrima) [inline security_monkey/scheduler.py]: Add region

File: security_monkey/scheduler.py (hunk 11 above)
Comment (mikegrima) [inline security_monkey/scheduler.py]: Move this inside of the loop.

File: security_monkey/datastore_utils.py [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+        return False, None, result
+
+
+def result_from_item(item, account, technology):
+    # Construct the query to obtain the specific item from the database:
+    return datastore.Item.query.filter(Item.name == item.name, Item.region == item.region,
+                                       Item.account_id == account.id, Item.tech_id == technology.id).scalar()
+
+
+def delete_old_revisions(watcher, arns, account, technology):
```
Comment (mikegrima) [inline security_monkey/datastore_utils.py]: Rename this to `inactivate_old_revisions`

File: security_monkey/datastore_utils.py [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+        config = {"Arn": db_item.arn}
+        revision = create_revision(config, db_item)
+        db_item.revisions.append(revision)
+
+        complete_hash, durable_hash = hash_item(config, watcher.ephemeral_paths)
+
+        db_item.latest_revision_complete_hash = complete_hash
+        db_item.latest_revision_durable_hash = durable_hash
+
+        db_item.latest_revision_id = revision.id
```
Comment (mikegrima) [inline security_monkey/datastore_utils.py]: Delete

File: security_monkey/datastore_utils.py [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+
+        # Do it again to update the latest revision ID:
+        datastore.db.session.refresh(revision)
+        db_item.latest_revision_id = revision.id
+        datastore.db.session.add(db_item)
+
+        # Find any audit issues associated with this revision, and delete it:
+        ia = ItemAudit.query.filter(ItemAudit.item_id == db_item.id).all()
+
+        for auditItem in ia:
```
Comment (mikegrima) [inline security_monkey/datastore_utils.py]: camelCase fix.

File: security_monkey/datastore_utils.py [hunk 18]
```diff
@@ -0,0 +1,248 @@
...
+        datastore.db.session.add(revision)
+        datastore.db.session.commit()
+
+        # Do it again to update the latest revision ID:
+        datastore.db.session.refresh(revision)
+        db_item.latest_revision_id = revision.id
+        datastore.db.session.add(db_item)
+
+        # Find any audit issues associated with this revision, and delete it:
+        ia = ItemAudit.query.filter(ItemAudit.item_id == db_item.id).all()
```
Comment (mikegrima) [inline security_monkey/datastore_utils.py]: See #564

File: security_monkey/scheduler.py [hunk 19]
```diff
@@ -102,6 +139,25 @@ def _audit_changes(account, auditors, send_report, debug=True):
         store_exception("scheduler-audit-changes", None, e)
 
 
+def _audit_specific_changes(monitor, audit_items, send_report, debug=True):
+    """ Runs auditors on all items """
```
Comment (mikegrima) [inline security_monkey/scheduler.py]: Fix this.

File: security_monkey/datastore.py [hunk 20]
```diff
@@ -714,8 +715,7 @@ def store_exception(source, location, exception, ttl=None):
                     db.session.add(technology)
                     db.session.commit()
                     db.session.refresh(technology)
-                    app.logger.info("Creating a new Technology: {} - ID: {}"
-                                    .format(technology.name, technology.id))
+                    app.logger.error("Creating a new Technology: {} - ID: {}".format(technology.name, technology.id))
```
Comment (mikegrima) [inline security_monkey/datastore.py]: Fixed.

File: security_monkey/datastore_utils.py (hunk 18 above)
Comment (mikegrima) [inline security_monkey/datastore_utils.py]: @monkeysecurity So shouldn't be relevant -- as deletion of audit issues only occur when the item itself is seen as deleted.

**Comment by mikegrima** [comment]
`python manage.py find_changes -m iamrole` breaks for some reason... Need to investigate.

File: security_monkey/tests/core/test_datastore_utils.py [hunk 21]
```diff
@@ -0,0 +1,343 @@
...
+    def __init__(self):
+        self.ephemeral_paths = []
+
+
+class DatabaseUtilsTestCase(SecurityMonkeyTestCase):
+    def pre_test_setup(self):
+        pass
+
+    def setup_db(self):
+        account_type_result = AccountType.query.filter(AccountType.name == 'AWS').first()
```
Comment (mikegrima) [inline security_monkey/tests/core/test_datastore_utils.py]: Fixed.

File: security_monkey/tests/core/test_watcher.py [hunk 22]
```diff
@@ -63,8 +121,22 @@ def test_from_items(self):
         assert len(merged_item_w_issues.audit_issues) == 1
         assert len(merged_item_wo_issues.audit_issues) == 0
 
-    def test_no_change_items(self):
+    def setup_batch_db(self):
+        account_type_result = AccountType.query.filter(AccountType.name == 'AWS').first()
+        if not account_type_result:
```
Comment (mikegrima) [inline security_monkey/tests/core/test_watcher.py]: Fixed.

**Comment by mikegrima** [comment]
More bugs to fix >:

Need to fix the `reporter.py`, since that apparently calls `slurp` directly:
```
        for monitor in mons:
            app.logger.info("Running slurp {} for {} ({} minutes interval)".format(monitor.watcher.i_am_singular, account, interval))
            (items, exception_map) = monitor.watcher.slurp()
            monitor.watcher.find_changes(items, exception_map)
            if (len(monitor.watcher.created_items) > 0) or (len(monitor.watcher.changed_items) > 0):
                watchers_with_changes.add(monitor.watcher.index)
            monitor.watcher.save()```

File: security_monkey/auditor.py [hunk 23]
```diff
@@ -165,14 +167,17 @@ def _is_current_method_disabled(self):
...
 
-
     def audit_all_objects(self):
         """
         Read all items from the database and inspect them all.
         """
         self.items = self.read_previous_items()
         self.audit_these_objects(self.items)
 
+    def audit_specific_items(self, items):
```
Comment (monkeysecurity) [inline security_monkey/auditor.py]: The names `audit_specific_items` and `audit_these_objects` seem to mean the same thing.

Why not delete `audit_specific_items` and have `audit_these_objects` always set `self.items = items`.

Comment (mikegrima) [inline security_monkey/auditor.py]: I have refactored this away. This did change logic in some other areas (refactored in latest push).  Some areas of the code need to `read_previous_items()` to get the items.

I have refactored this in the latest push where `auditor.items` must be set before running `auditor.audit_objects()`.

**Pull request merged**